src/rust/engine/process_execution/src/remote.rs

```diff
@@ -498,31 +507,27 @@ impl CommandRunner {
   }
 
   fn call_option(&self, build_id: String) -> Result<grpcio::CallOption, String> {
-    let mut builder = grpcio::MetadataBuilder::with_capacity(2);
-    if let Some(ref authorization_header) = self.authorization_header {
-      builder
-        .add_str("authorization", &authorization_header)
-        .unwrap();
-    }
-    {
-      let mut metadata = bazel_protos::remote_execution::RequestMetadata::new();
-      metadata.set_tool_details({
-        let mut tool_details = bazel_protos::remote_execution::ToolDetails::new();
-        tool_details.set_tool_name(String::from("pants"));
-        tool_details
-      });
-      metadata.set_tool_invocation_id(build_id);
-      // TODO: Maybe set action_id too
-      let bytes = metadata
-        .write_to_bytes()
-        .map_err(|err| format!("Error serializing request metadata proto bytes: {}", err))?;
-      builder
-        .add_bytes(
-          "google.devtools.remoteexecution.v1test.requestmetadata-bin",
-          &bytes,
-        )
-        .map_err(|err| format!("Error setting request metadata header: {}", err))?;
+    let mut builder = grpcio::MetadataBuilder::with_capacity(self.headers.len() + 1);
+    for (header_name, header_value) in &self.headers {
+      builder.add_str(header_name, header_value).unwrap();
```

Comment:
This unwrap will panic if the header is... illegal? That feels possible, and more likely if there are user-defined. Might be good to actually handle that.

Reply:
Also, the headers are defined per-CommandRunner, so roughly per-Core. Could it make sense to validate the headers in the constructor, so that we fail early? Or maybe even in python?

Reply:
Done